hw/dv/dpi/gpiodpi/gpiodpi.c

```diff
@@ -21,83 +21,238 @@
+         wfifo, n_bits);
+  printf("GPIO: To read the values of the pins relative to ground, run\n");
+  printf("$ cat %s  # '0' low, '1' high, 'X' floating\n", rfifo);
+  printf("GPIO: To drive the pins, run a command like\n");
+  printf("$ echo 'h0a l1f' > %s  # Pull the 10th pin high, and the 31st low.\n",
+         wfifo);
+}
+
 void *gpiodpi_create(const char *name, int n_bits) {
-  struct gpiodpi_ctx *ctx =
-      (struct gpiodpi_ctx *)calloc(1, sizeof(struct gpiodpi_ctx));
+  gpiodpi_ctx_t *ctx = (gpiodpi_ctx_t *)malloc(sizeof(gpiodpi_ctx_t));
   assert(ctx);
 
   // n_bits > 32 requires more sophisticated handling of svBitVecVal which we
   // currently don't do.
   assert(n_bits <= 32 && "n_bits must be <= 32");
-
   ctx->n_bits = n_bits;
 
-  char cwd[PATH_MAX];
-  char *cwd_rv;
-  cwd_rv = getcwd(cwd, sizeof(cwd));
-  assert(cwd_rv != NULL);
+  ctx->driven_pin_values = 0;
+
+  char cwd_buf[PATH_MAX];
+  char *cwd = getcwd(cwd_buf, sizeof(cwd_buf));
+  assert(cwd != NULL);
 
-  int rv;
-  rv = snprintf(ctx->fifo_pathname, PATH_MAX, "%s/%s", cwd, name);
-  assert(rv <= PATH_MAX && rv > 0);
+  int path_len;
+  path_len = snprintf(ctx->dev_to_host_path, PATH_MAX, "%s/%s-read", cwd, name);
+  assert(path_len > 0 && path_len <= PATH_MAX);
+  path_len =
+      snprintf(ctx->host_to_dev_path, PATH_MAX, "%s/%s-write", cwd, name);
+  assert(path_len > 0 && path_len <= PATH_MAX);
 
-  rv = mkfifo(ctx->fifo_pathname, 0644);  // writes are not supported currently
-  if (rv != 0) {
-    fprintf(stderr, "GPIO: Unable to create FIFO at %s: %s\n",
-            ctx->fifo_pathname, strerror(errno));
+  ctx->dev_to_host_fifo = open_fifo(ctx->dev_to_host_path, O_RDWR);
+  if (ctx->dev_to_host_fifo < 0) {
     return NULL;
   }
 
-  ctx->fifo_fd = open(ctx->fifo_pathname, O_RDWR);
-  if (ctx->fifo_fd < 0) {
-    fprintf(stderr, "GPIO: Unable to open FIFO at %s: %s\n", ctx->fifo_pathname,
-            strerror(errno));
+  ctx->host_to_dev_fifo = open_fifo(ctx->host_to_dev_path, O_RDWR);
+  if (ctx->host_to_dev_fifo < 0) {
     return NULL;
   }
 
-  printf(
-      "\n"
-      "GPIO: FIFO pipe created at %s for %d bit wide GPIO. Run\n"
-      "$ cat %s\n"
-      "to observe the output.\n",
-      ctx->fifo_pathname, ctx->n_bits, ctx->fifo_pathname);
+  int flags = fcntl(ctx->host_to_dev_fifo, F_GETFL, 0);
+  fcntl(ctx->host_to_dev_fifo, F_SETFL, flags | O_NONBLOCK);
+
+  print_usage(ctx->dev_to_host_path, ctx->host_to_dev_path, ctx->n_bits);
 
   return (void *)ctx;
 }
 
 void gpiodpi_device_to_host(void *ctx_void, svBitVecVal *gpio_data,
                             svBitVecVal *gpio_oe) {
-  struct gpiodpi_ctx *ctx = (struct gpiodpi_ctx *)ctx_void;
+  gpiodpi_ctx_t *ctx = (gpiodpi_ctx_t *)ctx_void;
   assert(ctx);
 
-  // n_bits > 32 requires more sophisticated handling of svBitVecVal which we
-  // currently don't do (loop through the array of 32 bit values).
-  assert(ctx->n_bits <= 32 && "n_bits must be <= 32");
```

Comment:
That's still valid, isn't it?

Reply:
No, this assert is run when the context is created, so we don't need it.